### ISSUE TITLE
Updating a materialized view may result in ghost rows appearing

### ISSUE THREAD

**Issue by psarna** [scylladb/scylla#5130]
This is Scylla's bug tracker, to be used for reporting bugs only.
If you have a question about Scylla, and not a bug, please ask it in
our mailing-list at scylladb-dev@googlegroups.com or in our slack channel.

- [x] I have read the disclaimer above, and I am reporting a suspected malfunction in Scylla.

*Installation details*
Scylla version (or git commit hash): master (c6b66d197bfb1114542749d7a1062eea756d44e7)

In current code, we start issuing view replica updates before calling `apply_with_commitlog`, which writes the data to the base replica. Consequently, if `apply_with_commitlog` fails, but some view replica updates managed to reach view replicas, the result is that the target materialized view contains rows that aren't represented by any row in the base table. Confirmed by programmatically injecting a failure into `apply_with_commitlog` and triggering a view update by writing to the base table.

Reported by @tgrabiec.

**Comment by nyh** [comment]
This is a known issue (but I don't remember if we ever documented it on the bug tracker), but only one of several MV consistency imperfections (it is not actually one of the scariest ones, because if the base update did reach additional base replicas, a base-table repair will fix this problem).

If we did it the other way around - i.e., only start sending the view update if the commit-log write succeeded, a failure at that point will result in the base table being updated but the view update will not, so is that better?

Also, even if we did do it in opposite order (commitog first), the data having been written to the commitlog doesn't guarantee anything - it still takes time for us to flush to commitlog, and even if we do, a node loss can result in the base copy being lost but the view copy not being lost.

In https://github.com/scylladb/scylla/wiki/Materialized-Views I explained how writes to a table with MV work in Cassandra (or at least worked, a few years ago when I wrote this), and you can see Cassandra did it the same order - it first started the remote view writes, and then wrote the base table locally.

If I remember correctly, I had this discussion once with @duarten (I can't find where) and we concluded that both orders are just as bad, so it's not very important which one we picked. Can you explain why you consider this order worse than the opposite order?

**Comment by psarna** [comment]
Yes, I'm aware that it's just one of the possible inconsistency scenarios, I created this issue simply as a follow-up of this discussion: https://groups.google.com/forum/#!topic/scylladb-dev/PU6TsY9RWNU%5B1-25%5D , so it's visible in our tracker. I don't claim that this order is worse than the opposite order at all, if Cassandra has the same one then at least we're "compatible" in that matter, I guess it's better than nothing :) Perhaps once we have base<->view repair, one of the scenarios will be easier to fix than the other, and then we can rethink swapping them.

**Comment by nyh** [comment]
Looking at the Cassandra code, I think I may have been wrong (either misunderstood it all along, or they changed it?) and it does use opposite order from us:

In `db/Keyspace.java`, `applyInternal()`, they first do
```java
getWriteHandler().beginWrite(mutation, makeDurable);
```
then
```java
viewManager.forTable(upd.metadata().id).pushViewReplicaUpdates(upd, makeDurable, baseComplete);
```
then finally
```
cfs.getWriteHandler().write(upd, ctx, indexTransaction);
```

It turns out that the `beginWrite()` (in `db/CassandraKeyspaceWriteHandler.java`) adds the mutation to the commitlog immediately. 

For us, this order is less natural: our `apply_with_commitlog()` writes to both commitlog and memtable, and we need to prepare the view update - requiring reading the table data before the memtable update - before updating the memtable. So to get the same order as Cassandra have, we should split apply_with_commitlog into separate commit and apply calls, do the commitlog first but the memtable apply only after preparing the view update.

**Comment by duarten** [comment]
This issue is contained in #3340 ("Bi becomes unavailable after updating the view table: the update becomes a ghost record in the view, absent from the base table"). View ghost rows are the hardest issues to deal with, and this is not the only way they can happen (another one is: you send the tombstone view update to a view, which then loses it). To solve them completely the mythical view <-> base repair mechanism would be needed. 

The reason this was coded this way was not only because it required less changes to the code (splitting the memtable and commitlog writes), but because writes are typically done with CL > 1, which means that eventually the base would be repaired to contain the new write.

**Comment by vladzcloudius** [comment]
@nyh @psarna I'm not sure I completely understand the last Duarte's comment related to what's needed in order to fix this.

Can't we simply send the corresponding MV replica update after the write to commitlog successfully resolves?

I this issue is still relevant I think we need to give this a priority. 

@nyh As for your comment that it would take longer to the view update to become visible - this is totally fine. 

Eventual consistence is ok. The complete lack of consistence - is not.

**Comment by psarna** [comment]
Without some kind of view<->base repair sending replicas after the local write will just swap one problem with another - for regular writes, instead of ghost rows there would be rows missing in the view, and in case of writing tombstones we can still end up with ghost rows (e.g. the tombstone is locally applied to commitlog first, but then it doesn't reach view replicas).

I think that Duarte's conclusion was that neither way completely solves the consistency problem, but writing to views first resulted in less code changes *and* base writes are more likely to get repaired (assuming CL>ONE), even if some local writes to commitlog failed.

**Comment by vladzcloudius** [comment]
@psarna What if write to the base fails? What if there is nothing to repair in the base? 

I agree that we need a view repair. However we don't have it and when we have it's going to be a very heavy operation. 

The regular base table repair is not going to fix the issue in question in cases when CQL Write to the base failed.

**Comment by vladzcloudius** [comment]
@eliransin ping.
What do you say about assigning this to somebody?

I still don't understand what's so complicated to schedule some action after the commitlog reports a success. We already have an action that takes place when this event occurs - we send a response to the coordinator. 
How hard it is to put another action there?

**Comment by tarzanek** [comment]
see https://forum.scylladb.com/t/scylladb-enterprise-release-2022-2-0/269#materialized-views-prune-5
it might be a workaround
